### ISSUE TITLE
Could ADBackend use GUID directly for filenames

### ISSUE THREAD

**Issue by willdean** [jakubgarfield/Bonobo-Git-Server#472]
The ADBackend currently makes names for json files by hashing the 'Name' property of the object.

Given that all objects now have GUIDs, and GUIDs make good filenames, could we now use the GUID directly rather than hashing the name?

In turn, the INameProperty could then become something like IHasGuidId, and only require objects to have a GUID, not a string name.

One for @larshg to answer, perhaps?

**Comment by RedX2501** [comment]
Before it was impossible to guarantee that username would not contain any weird properties that windows would not like. Now with GUIDs we have a unique way of identifying users in a filesystem friendly way. So i don't see why wo should not make use of that.

:+1: from me

**Comment by willdean** [comment]
Do all those backend files get rebuilt at startup/AD update?  i.e. if we change the filename format and hence effectively flush the whole store, is that a problem?

**Comment by RedX2501** [comment]
IIRC not at startup but with the background worker. When using AD there is no local configuration possible.

i also don't think there should be a problem with users no longer present in having any left over references in the system.

**Comment by larshg** [comment]
I don't think it should be a problem - but why bother?

Actually I can't really grasp why the ADBackend stores it info in files, rather than in a database to make the application more coherent.

**Comment by RedX2501** [comment]
@larshg I agree with you that it might not need a change but the advantage i see is that you could quickly identify a record in the backend that you are trying to access from the frontend.

**Comment by willdean** [comment]
I came to ask about the change because when I was changing the names of the properties on UserModel, I realised that the ADBackEnd needs something called 'Name' in INameProperty in order to generate a filename.

This made me realise that the ADBackEnd store will break if things are renamed, which is one of the the intentions of the GUID work.  Effectively the ADBackEnd store is primary-keyed on 'Name', where the EF stuff is now all keyed on the GUID Id.

Like @larshg, I'm skeptical about the ADBackend, particularly ADRepositoryRepository, which is not really doing anything that EFRepositoryRepository isn't doing, it's just doing it in a hand-rolled file database rather than Sqlite.  It's really easy to have subtle differences in routines like GetPermittedRepositories when they really need to be identical.

**Comment by larshg** [comment]
But @willdean if it is necessary for streamlining the AD part to be primarykeyed on GUID then I can't see any problems with it.

Its not a security risk either, as you can't really do anything with only the GUID of a given user?

**Comment by willdean** [comment]
@larshg I think it's going to be necessary to support renames.    Unfortunately I think we'll need some migration approach, because otherwise people will lose all the repository setup.

I have realised that we **can** unit-test _most_ of the AD code (including the backend store, but excluding the interaction with AD itself) - hopefully this would let me test the migration too.

I have just found another bug which is a subtle difference between ADRepositoryRepository and EFRepositoryRepository, so getting this stuff under the same set of tests would be a great bonus.
